File: web/stores/simulation.ts
import { create } from "zustand";
import type { TrainPosition, SimulationMetrics, SimulationState } from "@/types/api";

interface SimulationStore {
  trains: TrainPosition[];
  metrics: SimulationMetrics | null;
  state: SimulationState | null;
  tick: number;
  time_s: number;
  ist_time: string;
  service_period: string;
  depot_trains: number;
  completed_passengers: number;
  active_incidents: number;
  passengers: number;
  prev_completed: number;
  prev_time_s: number;
  setTrains: (trains: TrainPosition[]) => void;
  setMetrics: (m: SimulationMetrics) => void;
  setState: (s: SimulationState) => void;
  setTick: (tick: number, time_s: number) => void;
  setSimStats: (completed: number, incidents: number, passengers: number) => void;
  setServiceMeta: (istTime: string, period: string, depotTrains: number) => void;
}

export const useSimulationStore = create<SimulationStore>((set) => ({
  trains: [],
  metrics: null,
  state: null,
  tick: 0,
  time_s: 0,
  ist_time: "",
  service_period: "",
  depot_trains: 0,
  completed_passengers: 0,
  active_incidents: 0,
  passengers: 0,
  prev_completed: 0,
  prev_time_s: 0,
  setTrains: (trains) => set({ trains }),
  setMetrics: (metrics) => set({
    metrics: {
      avg_headway_s: metrics.avg_headway_s ?? 0,
      avg_dwell_s: metrics.avg_dwell_s ?? 0,
      avg_journey_time_s: metrics.avg_journey_time_s ?? 0,
      avg_speed_mps: metrics.avg_speed_mps ?? 0,
      total_energy_wh: metrics.total_energy_wh ?? 0,
    },
  }),
  setState: (state) => set({ state }),
  setTick: (tick, time_s) => set({ tick, time_s }),
  setSimStats: (completed, incidents, passengers) => set((s) => ({
    completed_passengers: completed,
    active_incidents: incidents,
    passengers,
    prev_completed: s.completed_passengers,
    prev_time_s: s.time_s,
  })),
  setServiceMeta: (ist_time, service_period, depot_trains) => set({ ist_time, service_period, depot_trains }),
}));

File: backend/app/schemas/simulation.py
from __future__ import annotations

from pydantic import BaseModel


class SimulationConfigSchema(BaseModel):
    duration_s: float = 3600.0
    dt_s: float = 1.0
    seed: int = 42
    n_passengers: int = 50000
    headway_target_s: float = 120.0
    snapshot_interval_s: float = 30.0


class SimulationState(BaseModel):
    running: bool
    paused: bool = False
    time_s: float
    trains: int
    active_trains: int
    depot_trains: int = 0
    passengers: int
    completed_passengers: int
    active_incidents: int
    service_period: str = ""
    ist_time: str = ""
    service_start: str = ""
    service_end: str = ""


class SimulationMetrics(BaseModel):
    avg_headway_s: float = 0.0
    avg_dwell_s: float = 0.0
    avg_journey_time_s: float = 0.0
    avg_speed_mps: float = 0.0
    total_energy_wh: float = 0.0


class TrainPosition(BaseModel):
    train_id: str
    line_code: str
    line_name: str = ""
    direction: str
    direction_destination: str = ""
    status: str
    speed_kmh: float
    speed_mps: float
    position_m: float
    current_station: str  # station code; use current_station_name for display
    current_station_name: str = ""
    next_station: str  # station code; use next_station_name for display
    next_station_name: str = ""
    occupancy: int
    doors_open: bool
    block_id: str


class DisruptRequest(BaseModel):
    station_code: str
    line_code: str = ""
    incident_type: str = "emergency_brake"
    duration_s: float = 300.0


class ApproachInfo(BaseModel):
    bearing: float
    line_code: str
    direction: str


class StationApproachData(BaseModel):
    id: str
    code: str
    name: str
    line_code: str
    latitude: float
    longitude: float
    platforms: int
    has_junction: bool


class ApproachingTrainsResponse(BaseModel):
    station_code: str
    station: StationApproachData
    trains: list[TrainPosition]
    approaches: list[ApproachInfo]
